### PR TITLE
Fix: convert Onnx DynamicSlice operator with 4 inputs to caffe2 fa…

### PR DESCRIPTION
**Pull request by hexiaoting**
I reported an issue [https://github.com/pytorch/pytorch/issues/20743](url)
and make this pull request for it

### CODE REVIEW

File: caffe2/onnx/backend.cc [hunk 1]
```diff
@@ -1082,7 +1082,7 @@ Caffe2Ops Caffe2Backend::CreateDynamicSlice(
   // arguments to the caffe2 Slice operator.
   std::string axes_tensor;
   if (onnx_node->node.input_size() > 2) {
-    axes_tensor = onnx_node->node.input(3);
+    axes_tensor = onnx_node->node.input(2);
```
Comment (BowenBao) [inline caffe2/onnx/backend.cc]: The DynamicSlice spec has 3-4 inputs, with the 4th optional input as ```axes```. So maybe the fix should be on the if condition
 ```if (onnx_node->node.input_size() > 3) {```

Comment (hexiaoting) [reply caffe2/onnx/backend.cc]: Oh, yes, DynamicSlice operator schema in ONNX has 3-4 inputs, the 4th is axes.
So I submit another commit, Is that OK?

**Review by houseroad — COMMENTED** [review]
Looks good. Could you add a test case to cover this code path?

Also please rebase to master to let the CI pass.

**Comment by hexiaoting** [comment]
@houseroad hi, I do not find where to add  a test case to cover this code path(convert onnx to caffe2 with specified operator), can you give me some tips or examples, Thank you.

**Comment by houseroad** [comment]
@hexiaoting I would suggest to add the test here: https://github.com/pytorch/pytorch/blob/master/test/onnx/test_pytorch_onnx_caffe2.py

**Comment by hexiaoting** [comment]
I have tried to add two test case to cover the function CreateDynamicSlice, But another problem come out when DynamicSlice operator with 3 inputs.
If dynamicslice operator have 3 inputs, then this function will create the following caffe2 operator
Shape
Shape
Range
...
In the situation, the second Shape's output is a 1D dimension tensor, But caffe2 Range operator only accept scalar Tensor whose dim() should be 0. This is a problem.
```
  auto* c2_op = ret.ops.Add();
  auto size_tensor = dummy_->NewDummyName();
  BuildOperator(c2_op, "Shape", {data}, {size_tensor});

  // Now get the rank of the tensor by getting the shape of the shape of
  // the input tensor
  c2_op = ret.ops.Add();
  auto rank_tensor = dummy_->NewDummyName();
  BuildOperator(c2_op, "Shape", {size_tensor}, {rank_tensor});

  // Axes tensor will be used to populate the fully-specified starts and ends
  // arguments to the caffe2 Slice operator.
  std::string axes_tensor;
  if (onnx_node->node.input_size() > 3) {                                                    ####test case1
    axes_tensor = onnx_node->node.input(3);
  } else {                                                                                                          #####test case2
    axes_tensor = dummy_->NewDummyName();
    auto* c2_op = ret.ops.Add();
    BuildOperator(c2_op, "Range", {rank_tensor}, {axes_tensor}, {});
  }
```

What I suggest to do for this problem  is accept 1 dimension with 1 element input for range.
How do you think? @houseroad

**Comment by houseroad** [comment]
Let's extend `Range` to support 1d tensor then :-)

**Comment by hexiaoting** [comment]
can I add a test case for dynamicSlice operator with 3 inputs that  will cover the codes I modified, and merge this PR first, then create another PR for range operator supporting 1D tensor ?

**Comment by hexiaoting** [comment]
1. add two testcase for dynamicSlice operator converted from onnx to caffe2 in file:http://test/onnx/test_pytorch_onnx_caffe2.py
2. extend range operator to support 1D tensor.
@houseroad

File: caffe2/operators/utility_ops.h [hunk 2]
```diff
@@ -1376,7 +1376,11 @@ class RangeOp : public Operator<Context> {
     T step = 1;
 
     for (int i = 0; i < InputSize(); ++i) {
-      CAFFE_ENFORCE_EQ(Input(0).dim(), 0, "All inputs must be scalar.");
+      if (Input(i).dim() == 1) {
```
Comment (houseroad) [inline caffe2/operators/utility_ops.h]: This doesn't need to 1, as long as numel() is 1 should be fine.

File: caffe2/operators/utility_ops.h [hunk 3]
```diff
@@ -1376,7 +1376,11 @@ class RangeOp : public Operator<Context> {
...
+        CAFFE_ENFORCE_EQ(Input(i).numel(), 1, "All inputs must be scalar/1D tensor.");
+      } else {
+        CAFFE_ENFORCE_EQ(Input(i).dim(), 0, "All inputs must be scalar.");
```
Comment (houseroad) [inline caffe2/operators/utility_ops.h]: update the error message here as well?

File: test/onnx/test_pytorch_onnx_caffe2.py [hunk 4]
```diff
@@ -1187,6 +1188,49 @@ def forward(self, x):
         x = torch.randn(3, 4, 5, 6, 7)
         self.run_model_test(NegSlice(), train=False, input=(x,), batch_size=BATCH_SIZE, use_gpu=False)
 
+    def test_dynamicslice_3inputs_onnx2caffe2(self):
```
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_caffe2.py]: I think https://github.com/pytorch/pytorch/blob/master/caffe2/python/onnx/tests/onnx_backend_test.py or https://github.com/pytorch/pytorch/blob/master/caffe2/python/onnx/tests/c2_ref_test.py is more appropriate place to add the following tests.

**Comment by hexiaoting** [comment]
@houseroad 
I change the position for DynamicSlice test case and modify the check code in caffe2/operators/utility_ops.h

File: caffe2/operators/utility_ops.cc [hunk 5]
```diff
@@ -1018,9 +1018,9 @@ output: [ 4  6  8 10 12 14 16]
     .Input(
         0,
         "start",
-        "(*Tensor*): [OPTIONAL] scalar tensor containing the start of the interval (inclusive) (default=0)")
-    .Input(1, "stop", "(*Tensor*): scalar tensor containing the end of the interval (exclusive)")
-    .Input(2, "step", "(*Tensor*): [OPTIONAL] scalar tensor specifying the spacing between values (default=1)")
+        "(*Tensor*): [OPTIONAL] scalar/1D tensor containing the start of the interval (inclusive) (default=0)")
```
Comment (houseroad) [inline caffe2/operators/utility_ops.cc]: scalar/1D => scalar or 1-element?

File: caffe2/operators/utility_ops.cc [hunk 6]
```diff
@@ -1018,9 +1018,9 @@ output: [ 4  6  8 10 12 14 16]
...
+    .Input(1, "stop", "(*Tensor*): scalar/1D tensor containing the end of the interval (exclusive)")
```
Comment (houseroad) [inline caffe2/operators/utility_ops.cc]: ditto

File: caffe2/operators/utility_ops.cc [hunk 7]
```diff
@@ -1018,9 +1018,9 @@ output: [ 4  6  8 10 12 14 16]
...
+    .Input(2, "step", "(*Tensor*): [OPTIONAL] scalar/1D tensor specifying the spacing between values (default=1)")
```
Comment (houseroad) [inline caffe2/operators/utility_ops.cc]: ditto

**Review by houseroad — COMMENTED** [review]
Almost there, some nits to address.

File: caffe2/operators/utility_ops.cc (hunk 7 above)
Comment (hexiaoting) [reply caffe2/operators/utility_ops.cc]: update the description, That's ok now.

File: caffe2/python/onnx/tests/c2_ref_test.py [hunk 8]
```diff
@@ -49,6 +49,46 @@ def test_check_arguments(self):
...
+             make_tensor_value_info("axes", onnx.TensorProto.INT32, (1, 2))],
+            outputs=[make_tensor_value_info("Y", onnx.TensorProto.FLOAT, (1, 2))])
+        model_def = make_model(graph_def, producer_name='caffe2-ref-test')
+        x = [[1,2,3,4],[5,6,7,8]]
+        start = [0, 1]
+        end = [4, 5]
+        axes = [1, 0]
+        prepared = c2.prepare(model_def)
+        caffe2_out = prepared.run(inputs=[np.array(x), np.array(start), np.array(end), np.array(axes)])
+        assert caffe2_out[0].shape == (1,4)
```
Comment (houseroad) [inline caffe2/python/onnx/tests/c2_ref_test.py]: I just realized that we don't compare the results with the reference implementation. Do you want to add this comparison?

File: caffe2/python/onnx/tests/c2_ref_test.py [hunk 9]
```diff
@@ -49,6 +49,46 @@ def test_check_arguments(self):
...
+             make_tensor_value_info("X3", onnx.TensorProto.INT32, (1, 2))],
+            outputs=[make_tensor_value_info("Y", onnx.TensorProto.FLOAT, (1, 2))])
+        model_def = make_model(graph_def, producer_name='caffe2-ref-test')
+
+        x = [[1,2,3,4],[5,6,7,8]]
+        start = [0, 0]
+        end = [-1, 4]
+        prepared = c2.prepare(model_def)
+        output = prepared.run(inputs=[np.array(x), np.array(start), np.array(end)])
+        assert output[0].shape == (1,4)
```
Comment (houseroad) [inline caffe2/python/onnx/tests/c2_ref_test.py]: same here. :-) Numpy slice should be enough for the reference implementation.

**Review by houseroad — APPROVED** [review]
Overall, looks good. Please address my inline comments. Then we are good to go.

File: caffe2/python/onnx/tests/c2_ref_test.py (hunk 9 above)
Comment (hexiaoting) [reply caffe2/python/onnx/tests/c2_ref_test.py]: Do you mean I use
```
self.assertSameOutputs(output[0], np.array(x)[0:-1, 0:4])
self.assertSameOutputs(output[0], np.array(x)[1:5, 0:4])
```
to instead line70 and line90 respectively?

**Comment by hexiaoting** [comment]
@houseroad why close this PR? It's not merged to pytorch repo right？

**Pull request closed without merging**